build(docs-infra): fail on unused doc-regions

petebacondarwin opened the pull request:
(I guess it is likely that this will not backport cleanly to patch)

Fixes #19761

petebacondarwin committed:
build(docs-infra): remove outdated comment

The comment referred to pre v9 Angular. The `http` package has
now been completely removed.

petebacondarwin committed:
build(docs-infra): improve docs-watch example matching

In the `docs-watch` job we attempt to guess what additional files need
to be processed given a change to a single file. Previously we were not
finding links to examples that were defined over multiple lines. E.g.

```
<code-pane
  header="src/app/app.component.html"
  region="hero-birthday-template"
  path="pipes/src/app/app.component.html">
</code-pane>
```

This commit improves the regular expression to handle these cases.

petebacondarwin committed:
build(docs-infra): fail if there are unused example regions

We can define regions in our examples that can be referenced
and rendered in guides as code snippets. It is quite hard to ensure
that these regions are maintained correctly. One reason for this is
it is hard to know whether a region is being used or not.

This commit adds a new processor that checks for unused named
regions in examples and fails if any are found.

Fixes #19761

petebacondarwin committed:
docs: remove unused "docregions"

This commit removes all the docregion tags in examples that are not
being referenced in any doc.

mary-poppins commented:
You can preview 4c5b4d1 at https://pr40479-4c5b4d1.ngbuilds.io/.
You can preview 5f581f9 at https://pr40479-5f581f9.ngbuilds.io/.

gkalpak commented on aio/content/examples/testing/src/app/demo/demo.testbed.spec.ts:
```diff
@@ -628,7 +617,7 @@ describe('demo (with TestBed):', () => {
       child.childValue = 'bar';
 
-      return new Promise(resolve => {
+      return new Promise<void>(resolve => {
```
This seems unrelated to this commit 🤔

petebacondarwin replied:
Well yes, but my IDE was saying that the file was invalid without this change.
So I thought I would just sneak it in :-D
I'll move it to its own commit.

gkalpak commented on packages/compiler/src/url_resolver.ts:
```diff
@@ -26,7 +26,7 @@ export function createOfflineCompileUrlResolver(): UrlResolver {
  * ## Example
  *
- * {@example compiler/ts/url_resolver/url_resolver.ts region='url_resolver'}
+ * <code-example path="compiler/ts/url_resolver/url_resolver.ts"></code-example>
```
Why remove the region here?

petebacondarwin replied:
See https://github.com/angular/angular/pull/40479#discussion_r560854939

gkalpak commented on packages/examples/compiler/ts/url_resolver/url_resolver.ts:
```diff
@@ -15,7 +15,6 @@ import {platformBrowserDynamic} from '@angular/platform-browser-dynamic';
 class MyApp {
 }
 
-// #docregion url_resolver
```
Why remove this docregion (since it was used in [packages/compiler/src/url_resolver.ts][1] before this PR)?

[1]: https://github.com/angular/angular/blob/0d51a81a44d58cbdcecdf73e60dc15c9e48ca364/packages/compiler/src/url_resolver.ts#L29

petebacondarwin replied:
My mistake!

petebacondarwin replied:
Actually, I did mean to do this.  The thing is that the `UrlResolver` in `packages/compiler/src/url_resolver.ts` is not public API. So the example there is not used in the docs. The result of this is that the doc-region here causes a failure, since it is not "used".

So I removed it. But then I thought, it doesn't make sense to reference the region in the non-public API doc, so I changed that to use the unnamed region, which would still work if that did become part of the public API.

gkalpak replied:
I see. That's not ideal, but if it's only one occurrence not worth trying to fix it "properly"(:tm:) :grin:

petebacondarwin replied:
It might be possible to avoid this problem by moving the `checkForUnusedExampleRegions` process earlier in the processing before we remove non-public APIs. But I have already spent more time than I had planned on this, so I am loathe to start further investigations.

gkalpak reviewed (changes requested):
:100: :100: :100:

petebacondarwin commented on aio/tools/transforms/authors-package/utils.js:
```diff
@@ -0,0 +1,11 @@
+ */
+
+module.exports = {
+  codeExampleMatcher: /<code-(?:pane|example)(?:[^>]|[ \r\n])*path="([^"]+)"/g,
```
The original regex did not match examples that covered multiple lines, e.g.

```
<code-example
  path="styleguide/src/05-03/app/heroes/shared/hero-button/hero-button.component.avoid.ts"
  region="example"
  header="app/heroes/hero-button/hero-button.component.ts">
</code-example>
```

I thought it was because we were not picking up the newlines, but now I see it was because the original regex required a space character after the element name. 🤦

petebacondarwin committed:
fixup! build(docs-infra): improve docs-watch example matching

petebacondarwin commented on aio/tools/transforms/angular-base-package/index.js:
```diff
@@ -53,10 +53,11 @@ module.exports = new Package('angular-base', [
+    // These are disabled here to prevent false negatives for the `docs-watch` task.
     // It is re-enabled in the main `angular.io-package`
     checkAnchorLinksProcessor.$enabled = false;
+    checkForUnusedExampleRegions.$enabled = false;
```
The trouble is that there are multiple authoring packages and we need to remember to disable it in each of them. So it felt more dry to do it this way round.
But perhaps we could refactor it so that all authoring packages extended a standard one that did these shared disablings.
I'll add an extra commit.

petebacondarwin committed:
fixup! build(docs-infra): remove outdated comment

petebacondarwin committed:
build(docs-infra): fail gently if a changed file is unknown

When using `docs-watch` we try to match the changed file
to a dgeni package in order to trigger a minimal set of processors.

Previously, if the changed file could not be matched we emitted an
ugly error. Now we provide a helpeful message.

petebacondarwin committed:
build(docs-infra): move docs-watch settings to a base package

This new `base-authoring-package` captures all the settings, which
turns of potentially fatal checks, in one place. This new package is
then used as a base for all the docs-watch related packages, rather
than dotting the settings in a variety of different packages. This also
has the benefit that the standard configuration for doc-gen is fatal
on failed checks by default.

petebacondarwin commented on aio/tools/transforms/examples-package/processors/check-for-unused-example-regions.spec.js:
```diff
@@ -0,0 +1,77 @@
+      expect(() => processor.$process())
+          .toThrowError('There is 1 unused example region:\n - some/file#not-used');
+    });
+  });
```
You assume correct!

petebacondarwin committed:
fixup! build(docs-infra): fail if there are unused example regions

petebacondarwin committed:
fixup! docs: remove unused "docregions"

petebacondarwin committed:
docs: add missing type parameter to testing example

Without this type parameter the call to `resolve()` is invalid.

mary-poppins commented:
You can preview 0fb7f09 at https://pr40479-0fb7f09.ngbuilds.io/.

petebacondarwin committed:
fixup! build(docs-infra): move docs-watch settings to a base package

petebacondarwin committed:
fixup! build(docs-infra): move docs-watch settings to a base package

mary-poppins commented:
You can preview b4ff749 at https://pr40479-b4ff749.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/authors-package/marketing-package.js:
```diff
@@ -7,10 +7,11 @@
  */
 const Package = require('dgeni').Package;
 const contentPackage = require('../angular-content-package');
+const baseAuthoringPackage = require('./base-authoring-package');
 const { CONTENTS_PATH } = require('../config');
```
😇 

```suggestion
const {CONTENTS_PATH} = require('../config');
const baseAuthoringPackage = require('./base-authoring-package');
```

gkalpak commented on aio/tools/transforms/authors-package/base-authoring-package.js:
```diff
@@ -0,0 +1,36 @@
+ */
+const baseAuthoringPackage = new Package('base-authoring', [basePackage]);
+baseAuthoringPackage
+    .config(function(checkAnchorLinksProcessor, checkForUnusedExampleRegions) {
+      // These are disabled here to prevent false negatives for the `docs-watch` task.
+      checkAnchorLinksProcessor.$enabled = false;
+      checkForUnusedExampleRegions.$enabled = false;
+    })
+
+    .config(function(linkInlineTagDef) {
+      // Do not fail the processing if there is an invalid link
+      linkInlineTagDef.failOnBadLink = false;
+    })
+
+    .config(function(renderExamples) {
+      // Do not fail the processing if there is a broken example
+      renderExamples.ignoreBrokenExamples = true;
+    });
```
OOC, why the separate `config()` blocks for `linkInlineTagDef` and `renderExamples`, but not for `checkAnchorLinksProcessor` and `checkForUnusedExampleRegions`?

petebacondarwin replied:
Generally I like to keep the config blocks focussed on a single task. In this case the first block is disabling two processors. In the other blocks we are configuring different aspects of other injectables. They could indeed all go in one, but I have found that this leads to people bundling lots of stuff into a single block, so I like to start out as I wish to carry on. Equally the first config block could have been split into two but then I would have had to duplicate the comment :-D

gkalpak commented on aio/tools/transforms/angular-base-package/index.js:
```diff
@@ -5,17 +5,21 @@
+const { SRC_PATH } = require('../config');
 const linksPackage = require('../links-package');
 const examplesPackage = require('../examples-package');
 const targetPackage = require('../target-package');
 const remarkPackage = require('../remark-package');
```
Since you're ordering imports alphabetically :grin:

```suggestion
const remarkPackage = require('../remark-package');
const targetPackage = require('../target-package');
```

mary-poppins commented:
You can preview 7aab5f6 at https://pr40479-7aab5f6.ngbuilds.io/.

gkalpak reviewed (approved):
Great work :tada: I love the clean-up :heart:
(All these unused docregions have been bugging me for a long time :grin:)

gkalpak reviewed (approved):
Reviewed-for: global-docs-approvers

petebacondarwin commented:
Caretaker: This PR is a docs/docs-infra only change so a presubmit should not be required.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

Note: from `public-api` perspective there are just 3 files in this folder: `aio/content/examples/styleguide`.

jessicajaniuk reviewed (approved):
reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api
